### PR TITLE
[Generic] Implement refresh_user to periodically refresh OAuth tokens

### PR DESCRIPTION
**Pull request by jthiltges**
An attempt at OAuth token refresh, following the pattern in PR https://github.com/jupyterhub/oauthenticator/pull/287, without relying on tokens to be JWTs. This should address at least some of issue https://github.com/jupyterhub/oauthenticator/issues/398.

Requirements for token refresh:

- OAuth server response includes `expires_in` and `refresh_token`
- State is persisted on hub with `c.Authenticator.enable_auth_state = True` (to store the `refresh_token`)

### CODE REVIEW

**Comment by kkaraivanov1** [comment]
As in the next release of jupyterhub there will be authenticator user group management https://github.com/jupyterhub/jupyterhub/pull/3548 what happens if we have manage_groups=true  and the user's group membership changes after auth and before refresh? Your refresh code is not updating user_info['groups'] as it seem.

**Comment by jthiltges** [comment]
@kkaraivanov1 Thanks for looking at the PR. I agree that group data should be refreshed as well.

Is group membership info already being stored on authentication?

**Comment by kkaraivanov1** [comment]
The way I understand the intentions behind the PR I mention is  if c.Authenticator.manage_groups is set AND if the GenericOAuthenticator adds the groups to user_info['groups']  they will be populated automatically. 
So to answer your question - it could be 
Perhaps you can hide that behind a check like `if hasattr(self,manage_groups):  .......`. Please take my thoughts with big grain of salt, I'm not a dev and my python is rudimentary.

**Comment by jthiltges** [comment]
That makes sense, and thank you for the clarification.

Adding support for `user_info['groups']` is probably best addressed in a separate PR. I'll be happy to revise this PR to follow the pattern that the project maintainers prefer.

File: oauthenticator/generic.py [hunk 1]
```diff
@@ -216,6 +224,46 @@ async def authenticate(self, handler, data=None):
...
+
+        # Attempt renewal
+        params = dict(
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            refresh_token=refresh_token,
+            grant_type='refresh_token',
+        )
+
+        headers = self._get_headers()
```
Comment (kkaraivanov1) [inline oauthenticator/generic.py:256]: this one is wrong. It has to be 
`headers = { "Content-Type": "application/x-www-form-urlencoded" }`

**Pull request closed without merging**